full eslint

tharvik opened the pull request:
found out that eslint was not working on much files (ie, only on javascript files that we don't have anymore).

* commonize the eslint configuration to process all typescript files around
* fix a bunch of lint issues
* also making discojs-core platform independant, by using isomorphic-* libs and ES standards
  * upstream isomorphic-wrtc [is answering slowly](https://github.com/patagona-technologies/isomorphic-wrtc/pull/1), keeping an updated version around until then

tharvik committed:
discojs-core: use isomorphic-wrtc

tharvik committed:
discojs-core: drop node:url

tharvik committed:
*: commonize eslint

tharvik committed:
mocha: add ts-node dep in consumers

JulienVig reviewed (changes requested):
Very nice! 

It seems `node-webrtc` doesn't support Apple Silicon, indeed on M2 running `npm ci` raises this error:
```
npm ERR! code 1
npm ERR! path /Users/vignoud/Disco/review_repo/node_modules/wrtc
npm ERR! command failed
npm ERR! command sh -c -- node scripts/download-prebuilt.js
npm ERR! response status 404 Not Found on https://node-webrtc.s3.amazonaws.com/wrtc/v0.4.7/Release/darwin-arm64.tar.gz
npm ERR! node-pre-gyp info it worked if it ends with ok
npm ERR! node-pre-gyp info using node-pre-gyp@1.0.11
npm ERR! node-pre-gyp info using node@16.20.2 | darwin | arm64
npm ERR! node-pre-gyp info check checked for "/Users/vignoud/Disco/review_repo/node_modules/wrtc/build/Release/wrtc.node" (not found)
npm ERR! node-pre-gyp http GET https://node-webrtc.s3.amazonaws.com/wrtc/v0.4.7/Release/darwin-arm64.tar.gz
npm ERR! node-pre-gyp ERR! install response status 404 Not Found on https://node-webrtc.s3.amazonaws.com/wrtc/v0.4.7/Release/darwin-arm64.tar.gz
npm ERR! node-pre-gyp ERR! install error
npm ERR! node-pre-gyp ERR! stack Error: response status 404 Not Found on https://node-webrtc.s3.amazonaws.com/wrtc/v0.4.7/Release/darwin-arm64.tar.gz
npm ERR! node-pre-gyp ERR! stack     at /Users/vignoud/Disco/review_repo/node_modules/@mapbox/node-pre-gyp/lib/install.js:67:15
npm ERR! node-pre-gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:96:5)
npm ERR! node-pre-gyp ERR! System Darwin 23.3.0
npm ERR! node-pre-gyp ERR! command "/Users/vignoud/.nvm/versions/node/v16.20.2/bin/node" "/Users/vignoud/Disco/review_repo/node_modules/.bin/node-pre-gyp" "install"
npm ERR! node-pre-gyp ERR! cwd /Users/vignoud/Disco/review_repo/node_modules/wrtc
npm ERR! node-pre-gyp ERR! node -v v16.20.2
npm ERR! node-pre-gyp ERR! node-pre-gyp -v v1.0.11
npm ERR! node-pre-gyp ERR! not ok
```

I found a few open issues with a similar installation error:
- [#680](https://github.com/node-webrtc/node-webrtc/issues/680)
- [#698](https://github.com/node-webrtc/node-webrtc/issues/698)
- [#725](https://github.com/node-webrtc/node-webrtc/issues/725)
- [#729](https://github.com/node-webrtc/node-webrtc/issues/729)
- others I didn't check

None of the fixes I tried worked on M2. I didn't try [this one](https://github.com/node-webrtc/node-webrtc/issues/698#issuecomment-941861645).

FYI issue #698 mentions using `@koush/wrtc` instead, which I see you removed from the dependencies. I tried `npm i @koush/wrtc` anyway but still doesn't work so I don't know what changed compared to when it used to work on my M2.

tharvik commented:
> It seems `node-webrtc` doesn't support Apple Silicon
> [...]
> FYI issue #698 mentions using `@koush/wrtc` instead, which I see you removed from the dependencies. I tried `npm i @koush/wrtc` anyway but still doesn't work so I don't know what changed compared to when it used to work on my M2.

ho it remaines me that we add that, I though it was an issue of early userspace M2 support and that was fixed in the meantime. looks like it wasn't.

`wrtc` is an abandonned project, a bunch of people are actually building it by themself, publishing binaries around but there isn't a concerted effort. it'll be nice to actually get rid of it a some point and use node-datachannel that seems more modern. tracking in #633.
~~for quick fixes, `@koush/wrtc` was last released 2y ago, I'm more in favor of using `@roamhq/wrtc` that got released some months ago (and is the only one that updated its readme expliciting support for node-18)~~ we are still at node-16, I mixed my PRs oupsi

tharvik committed:
iso-wrtc: use wrtc compatible with Apple M2

tharvik commented:
@JulienVig as I don't have a M2 around, can you try out the latest commit plz? :)

JulienVig commented on discojs/discojs-core/src/async_informant.ts:
```diff
@@ -1,4 +1,4 @@
-import { AggregatorBase } from './aggregator'
+import { type AggregatorBase } from './aggregator'
```
I see you're adding type to imports everywhere, what changed?

tharvik replied:
eslint asked me to and I happily obliged :bow:
it's [a new syntax](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) to help bundlers by avoiding any reference to types that are not instanciated

JulienVig commented on discojs/discojs-core/src/client/event_connection.ts:
```diff
@@ -117,17 +114,18 @@ export class WebSocketServer implements EventConnection {
 
       // Validate message format
-      if (validateReceived && !validateReceived(msg)) {
+      if (validateReceived !== undefined && !validateReceived(msg)) {
```
Why is it better than testing `validateReceived`directly? Is it for legibility?

tharvik replied:
yeah, it's more explicit to actively check for undef rather than letting the truthy magic of JS operate

JulienVig commented on discojs/discojs-core/src/dataset/data/image_data.ts:
```diff
@@ -18,7 +18,7 @@ export class ImageData extends Data {
     // cause an error during training, because of the lazy aspect of the dataset; we only
     // verify the first sample.
-    if (!task.trainingInformation.preprocessingFunctions?.includes(ImagePreprocessing.Resize)) {
+    if (task.trainingInformation.preprocessingFunctions?.includes(ImagePreprocessing.Resize) !== true) {
```
Why is it better this way?

tharvik replied:
I agree, it's quite weird to test equality with true, hurt me to write it, but eslint doesn't want nullish test, only explicit ones. but as I'm rereading it, it's doesn't improve readability at all.
WDYT of the next one?
```ts
const hasResize = (
  task.trainingInformation.preprocessingFunctions ?? []
).includes(ImagePreprocessing.Resize);
if (!hasResize) { // ..
```

JulienVig replied:
I find the last suggestion clearer but I'll let you choose how want to do it!

JulienVig commented on discojs/discojs-core/src/dataset/data/preprocessing/tabular_preprocessing.ts:
```diff
@@ -20,13 +20,13 @@ const sanitize: PreprocessingFunction = {
       return {
-        xs: xs.map(i => i === undefined ? 0 : i),
-        ys: ys
+        xs: xs.map(i => i ?? 0),
+        ys
```
Oh very nice, I'll remember that.
Can you return `ys` directly because it has the same name as the key?

tharvik replied:
yes, it is indeed way nicer! that's the [property shorthand](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#property_definitions)

JulienVig commented on web-client/src/components/data/dataset_input/DatasetInput.vue:
```diff
@@ -166,7 +166,7 @@ const requireLabels = computed(
 
 const cannotUploadFiles = computed(
-  () => requireLabels && !csvRows.value
+  () => requireLabels.value && !csvRows.value
```
Good catch!

tharvik replied:
that's why we need linters :)

JulienVig reviewed (dismissed):
Very nice, the code is much cleaner!

I saw you created a new module `isomorphic-wrtc`, could you maybe add a few lines of explanation in the doc about what's the purpose of this module? For example as a README.md in the `isomorphic-wrtc` or in the Structure section of DEV.md

JulienVig reviewed (changes requested):
`npm -ws run build` fails because the isomorphic-wrtc doesn't have a build script:

> Lifecycle script `build` failed with error:
> npm ERR! Error: Missing script: "build"

tharvik committed:
iso-wrtc: add readme

tharvik committed:
iso-wrtc: add empty scripts

tharvik commented:
> I saw you created a new module `isomorphic-wrtc`, could you maybe add a few lines of explanation in the doc about what's the purpose of this module?

indeed, it's a bit cryptic when you first come to the project. added some infos in isomorphic-wrtc/README.md

> `npm -ws run build` fails because the isomorphic-wrtc doesn't have a build script:

ho right, I didn't check the return code when doing that. added empty scripts for build/lint/test

JulienVig reviewed (approved):
Looks good to me!
